### PR TITLE
RD-3017 Add support for switching between Catalog and Table view in Blueprints Marketplace modal.

### PR DESCRIPTION
**Pull request by stepek**
## Description
Add possibility to customize display style of blueprint catalog widget.

## Screenshots / Videos
image

## Checklist

- [X] My code follows the UI Code Style.
- [x] I verified that all tests and checks have passed.
- [X] I verified if that change requires a documentation update.
- [X] I added proper labels to this PR.

## Tests
[Build #1191 (03-Sep-2021 11:10:38)](https://jenkins.cloudify.co/view/UI%20Health%20View/job/Stage-UI-System-Test/1191/)
[Build #1233 (9 Sep 2021, 11:00:55)](https://jenkins.cloudify.co/view/UI%20Health%20View/job/Stage-UI-System-Test/1233/)
[ Build #1235 (9 Sep 2021, 11:36:38)](https://jenkins.cloudify.co/view/UI%20Health%20View/job/Stage-UI-System-Test/1235/)

 

## Documentation
Will add in separate ticket.

### CODE REVIEW

File: test/cypress/integration/widgets/blueprint_catalog_spec.ts [hunk 1]
```diff
@@ -34,4 +36,20 @@ describe('Blueprints catalog widget', () => {
         cy.contains('.segment', blueprintName).contains('Upload').click();
         cy.contains('.ui.label.section.active.pageTitle', blueprintName);
     });
+
+    it('should allow to change display style', () => {
+        cy.setDropdownConfigurationField('blueprintCatalog', 'Display style', ['Table']);
+        cy.get('table').should('be.visible');
```
Comment (kubama) [inline test/cypress/integration/widgets/blueprint_catalog_spec.ts:42]: I think we should be more specific here, at least `.blueprintCatalog table`.

File: test/cypress/integration/widgets/blueprint_catalog_spec.ts [hunk 2]
```diff
@@ -34,4 +36,20 @@ describe('Blueprints catalog widget', () => {
...
+
+    it('should allow to customize fields to show', () => {
+        cy.setDropdownConfigurationField('blueprintCatalog', 'List of fields to show in the table', [
+            'Name',
+            'Created'
+        ]);
+        cy.contains('Name').should('be.visible');
+        cy.contains('Created').should('be.visible');
+        cy.contains('Updated').should('not.exist');
+        cy.contains('Description').should('not.exist');
```
Comment (kubama) [inline test/cypress/integration/widgets/blueprint_catalog_spec.ts:53]: Similarly here, I'd use more specific selectors.

File: test/cypress/support/commands.ts [hunk 3]
```diff
@@ -404,6 +403,14 @@ const commands = {
             .click();
     },
 
+    clearDropdown: (fieldName: string) => {
+        cy.contains('.field', fieldName).within(() => {
+            if (Cypress.$(`.field:contains('${fieldName}') .delete.icon`).length > 0) {
+                cy.get('.delete.icon').click({ multiple: true });
+            }
```
Comment (kubama) [inline test/cypress/support/commands.ts:410]: What if the the `if` condition is not met? The command will just silently do nothing? Shouldn't it fail instead as all built-in Cypress commands do?
Also, why the `multiple` flag?

File: widgets/blueprintCatalog/src/widget.tsx [hunk 4]
```diff
@@ -59,9 +59,17 @@ Stage.defineWidget<
...
             type: Stage.Basic.GenericField.LIST_TYPE
         },
+        {
+            id: 'fieldsToShow',
+            name: 'List of fields to show in the table',
+            placeHolder: 'Select fields from the list',
+            items: ['Name', 'Description', 'Created', 'Updated'],
+            default: 'Name,Description,Created,Updated',
+            type: Stage.Basic.GenericField.MULTI_SELECT_LIST_TYPE
+        },
```
Comment (kubama) [inline widgets/blueprintCatalog/src/widget.tsx:72]: Why no i18n?

File: widgets/blueprints/src/widget.tsx [hunk 5]
```diff
@@ -48,7 +48,31 @@ Stage.defineWidget<unknown, unknown, BlueprintsWidgetConfiguration>({
...
+            ],
+            default: 'table',
+            type: Stage.Basic.GenericField.LIST_TYPE
+        },
+        {
+            id: 'marketplaceColumnsToShow',
+            name: Stage.i18n.t('widgets.blueprints.configuration.marketplaceColumnsToShow.label'),
+            placeHolder: Stage.i18n.t('widgets.blueprints.configuration.marketplaceColumnsToShow.placeHolder'),
+            items: ['Name', 'Description', 'Created', 'Updated'],
+            default: 'Name,Description',
```
Comment (kubama) [inline widgets/blueprints/src/widget.tsx:73]: I guess these could be also i18nized

File: widgets/blueprintCatalog/src/RepositoryTable.tsx [hunk 6]
```diff
@@ -31,10 +32,10 @@ const RepositoryTable: FunctionComponent<RepositoryViewProps> = ({
...
             noDataMessage={noDataMessage}
         >
-            <DataTable.Column label="Name" width="25%" />
-            <DataTable.Column label="Description" width="40%" />
-            <DataTable.Column label="Created" width="12%" />
-            <DataTable.Column label="Updated" width="12%" />
+            <DataTable.Column label="Name" width="25%" show={fieldsToShow.includes('Name')} />
+            <DataTable.Column label="Description" width="40%" show={fieldsToShow.includes('Description')} />
+            <DataTable.Column label="Created" width="12%" show={fieldsToShow.includes('Created')} />
+            <DataTable.Column label="Updated" width="12%" show={fieldsToShow.includes('Updated')} />
```
Comment (kubama) [inline widgets/blueprintCatalog/src/RepositoryTable.tsx:38]: Another place for i18nization.

Idea for future improvement: extend `DataTable` so it handled `fieldsToShow` internally.

File: test/cypress/support/commands.ts (hunk 3 above)
Comment (stepek) [reply test/cypress/support/commands.ts:410]: This function should clear dropdown when there is any item selected, so when it is clean we can't clear it again

`multiple` flag is for clicking multiple founded items by `get` command.

File: test/cypress/support/editMode.ts [hunk 7]
```diff
@@ -64,6 +64,12 @@ const commands = {
     setSearchableDropdownConfigurationField: (widgetId: string, fieldName: string, value: string) =>
         cy.editWidgetConfiguration(widgetId, () => {
             cy.setSearchableDropdownValue(fieldName, value);
+        }),
+
+    setDropdownConfigurationField: (widgetId: string, fieldName: string, value: string[]) =>
```
Comment (qooban) [inline test/cypress/support/editMode.ts:69]: 1. I think more accurate would be to have 
```suggestion
    setDropdownConfigurationField: (widgetId: string, fieldName: string, values: string[]) =>
```
2. You use it for multiple dropdown, but it will work also for single selection dropdowns, right?
3. Maybe `clearDropdown` should be done inside `setDropdownValues`?

File: test/cypress/support/commands.ts [hunk 8]
```diff
@@ -394,8 +394,7 @@ const commands = {
         }
     },
 
-    clearSearchableDropdown: (fieldName: string) =>
-        cy.contains('.field', fieldName).find('.dropdown.clear.icon').click(),
+    clearSearchableDropdown: (fieldName: string) => cy.contains('.field', fieldName).find('.clear.icon').click(),
```
Comment (qooban) [inline test/cypress/support/commands.ts:397]: Unrelated change or intentional?

File: test/cypress/support/commands.ts (hunk 3 above)
Comment (qooban) [reply test/cypress/support/commands.ts:410]: Is it meant to work with single selection dropdown (without `multiple` prop set) as well?
If not, I think we should name it `clearMultipleDropdown` or something like that to be more descriptive about the command purpopse.

File: widgets/common/src/blueprintMarketplace/BlueprintMarketplaceModal.tsx [hunk 9]
```diff
@@ -3,13 +3,15 @@ import { MarketplaceTab } from './types';
 
 interface BlueprintMarketplaceModalProps {
     open: boolean;
-    tabs: MarketplaceTab[];
+    tabs?: MarketplaceTab[];
+    displayStyle?: 'catalog' | 'table';
```
Comment (qooban) [inline widgets/common/src/blueprintMarketplace/BlueprintMarketplaceModal.tsx:7]: Seems like you defined the same union type few times. Maybe better would be to extract it to a reusable type? 
Maybe to `widgets/common/src/blueprintMarketplace/types.ts` and export it to `Stage.Common` similarly to how `MarketplaceTab` is exported?

File: widgets/common/src/blueprintMarketplace/BlueprintMarketplaceModal.tsx [hunk 10]
```diff
@@ -3,13 +3,15 @@ import { MarketplaceTab } from './types';
...
+    columns?: string[];
     onHide: () => void;
 }
 
 const t = Stage.Utils.getT('widgets.common.blueprintMarketplace');
 
-const getPageLayout = (tabs: MarketplaceTab[]) => ({
+const getPageLayout = (tabs: MarketplaceTab[], displayStyle: string, columns: string[]) => ({
```
Comment (qooban) [inline widgets/common/src/blueprintMarketplace/BlueprintMarketplaceModal.tsx:14]: We will have conflict here (see changes in #1576 ), but should be easy to resolve.

File: widgets/blueprints/src/widget.tsx [hunk 11]
```diff
@@ -48,7 +48,31 @@ Stage.defineWidget<unknown, unknown, BlueprintsWidgetConfiguration>({
             name: Stage.i18n.t('widgets.blueprints.configuration.showComposerOptions'),
             default: false
         },
-        Stage.Common.BlueprintMarketplace.tabsConfig
+        Stage.Common.BlueprintMarketplace.tabsConfig,
+        {
+            id: 'marketplaceDisplayStyle',
+            name: Stage.i18n.t('widgets.blueprints.configuration.marketplaceDisplayStyle.label'),
```
Comment (qooban) [inline widgets/blueprints/src/widget.tsx:54]: Lines are getting quite long. What do you think about having `t`-function here?
```
const t = Stage.Utils.getT(`widgets.blueprints...`);
```

In general:
1. Worth to check if the whole file can be translated.
2. Worth to check how difficult would it be to have full TS migration.

File: widgets/blueprints/src/types.ts [hunk 12]
```diff
@@ -8,6 +8,8 @@ export interface BlueprintsWidgetConfiguration {
     hideFailedBlueprints: boolean;
     showComposerOptions: boolean;
     marketplaceTabs: Stage.Common.BlueprintMarketplace.Tab[];
+    marketplaceDisplayStyle: 'table' | 'catalog';
```
Comment (qooban) [inline widgets/blueprints/src/types.ts:11]: Maybe we can have something like
```suggestion
    marketplaceDisplayStyle: Stage.Common.BlueprintMarketplace.DisplayStyle;
```
?
(see my comment about extracting a type for `'table' | 'catalog'`)

File: widgets/blueprintCatalog/src/widget.tsx [hunk 13]
```diff
@@ -59,9 +59,17 @@ Stage.defineWidget<
                 { name: 'Table', value: 'table' },
                 { name: 'Catalog', value: 'catalog' }
             ],
-            default: 'table',
+            default: 'catalog',
```
Comment (qooban) [inline widgets/blueprintCatalog/src/widget.tsx:62]: This change was intentional?

**Review by qooban — COMMENTED** [review]
Added some comments.

Are you planning to change display style in "Service Button" widget within this ticket or should I create another one?

File: test/cypress/support/commands.ts (hunk 8 above)
Comment (stepek) [reply test/cypress/support/commands.ts:397]: Unrelated change, sorry about that.

File: test/cypress/support/editMode.ts (hunk 7 above)
Comment (stepek) [reply test/cypress/support/editMode.ts:69]: 1. fixed 
2. should work
3. no, imho having it separately, keep current implementation of `set Dropdown Values` does not change current behavior of this method

File: test/cypress/support/commands.ts (hunk 3 above)
Comment (kubama) [reply test/cypress/support/commands.ts:411]: I'm still not sure about the implementation of the command. If a command is intended to do something then I think it should either do it or fail if it's not possible to perform the action. It should also wait for a state change if it's not possible to perform the action at the time the command is executed. This implementation breaks this contract, thus:
- it's not in line with other Cypress commands, so it's behavior is unexpected
- it's error prone - since there is no wait for proper state (availability of delete icons) the result of command execution depends on exact timing
- it makes tests less robust - there are two possible execution paths (the `if`) while in any particular case only one is valid - in a robust test the command should be called in cases where there are items to be cleared, it shouldn't be called at all in cases when there is nothing selected

File: test/cypress/support/editMode.ts [hunk 14]
```diff
@@ -64,6 +64,12 @@ const commands = {
     setSearchableDropdownConfigurationField: (widgetId: string, fieldName: string, value: string) =>
         cy.editWidgetConfiguration(widgetId, () => {
             cy.setSearchableDropdownValue(fieldName, value);
+        }),
+
+    setDropdownConfigurationField: (widgetId: string, fieldName: string, values: string[]) =>
+        cy.editWidgetConfiguration(widgetId, () => {
+            cy.clearMultipleDropdown(fieldName);
+            cy.setDropdownValues(fieldName, values);
```
Comment (qooban) [inline test/cypress/support/editMode.ts:72]: I see you changed `clearDropdown` into `clearMultipleDropdown`, so does it mean it was meant to work only with `Dropdown` component having `multiple` prop on? (as I have asked in https://github.com/cloudify-cosmo/cloudify-stage/pull/1581/files#r701743728)

If so, then I think than now `setDropdownConfigurationField` name is misleading. It shouldn't be used for single selection dropdowns, right?

File: widgets/blueprints/src/types.ts [hunk 15]
```diff
@@ -8,6 +8,8 @@ export interface BlueprintsWidgetConfiguration {
     hideFailedBlueprints: boolean;
     showComposerOptions: boolean;
     marketplaceTabs: Stage.Common.BlueprintMarketplace.Tab[];
+    marketplaceDisplayStyle: Stage.Common.BlueprintMarketplace.DisplayStyle;
+    marketplaceColumnsToShow: string[];
```
[the remark this answers is not in the document]
Comment (stepek) [reply widgets/blueprints/src/types.ts:12]: are you want add typings for marketplaceColumnsToShow based on translation file?

[the remark this answers is not in the document]
Comment (qooban) [reply widgets/blueprints/src/types.ts:12]: Hmmm... I'm not sure if that would be the best solution. We would need to make strict connection between `types.ts` and `en.json` file and that is probably something which we don't want to do. 

Maybe just strings union would be sufficient? Will it be valuable or am I wrong?

File: test/cypress/support/commands.ts (hunk 3 above)
Comment (stepek) [reply test/cypress/support/commands.ts:411]: Ok @kubama let me remove this if statement.

File: widgets/blueprints/src/types.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (stepek) [reply widgets/blueprints/src/types.ts:12]: no, it's not good solution, maybe the column keys should be const and should be separated from translations?

[the remark this answers is not in the document]
Comment (qooban) [reply widgets/blueprints/src/types.ts:12]: Seeing that it's not trivial and we don't have unlimited time for 6.2 features, let's maybe keep `string[]` for now.

**Review by qooban — APPROVED** [review]
I'm OK with the code right now.
Let's wait for @kubama review.

**Comment by qooban** [comment]
@stepek FYI There are conflicts to be resolved.

File: test/cypress/support/commands.ts [hunk 16]
```diff
@@ -375,42 +375,38 @@ const commands = {
...
-        cy.contains('.field', fieldName).within(() => {
-            if (Cypress.$(`.field:contains('${fieldName}') .delete.icon`).length > 0) {
-                cy.get('.delete.icon').click({ multiple: true });
-            }
-        });
-    },
+    clearMultipleDropdown: (fieldName: string) =>
+        cy.getField(fieldName).within(() => {
+            cy.get('.delete.icon').click({ multiple: true });
+        }),
```
Comment (kubama) [inline test/cypress/support/commands.ts:409]:
```suggestion
        cy.getField(fieldName).find('.delete.icon').click({ multiple: true });
```

**Review by kubama — APPROVED** [review]

**Review by qooban — APPROVED** [review]

**Comment by stepek** [comment]
https://jenkins.cloudify.co/view/UI%20Health%20View/job/Stage-UI-System-Test/1235/

**Review by kubama — APPROVED** [review]

**Review by qooban — APPROVED** [review]

**Comment by stepek** [comment]
Tests passing
